// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package abesim;

import "agent_based_epidemic_sim/core/pandemic.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";

// This message is used by downstream code for training the risk score model.
//
// This message contains a repeated field of ExposureResults that represent the
// test result for a given agent along with all of the exposures (confirmed and
// unconfirmed) in the temporal window defined by exposure_start_time and
// exposure_end_time. In general, one output message will be written per
// timestep in the eval.
message ExposuresPerTestResult {
  // In this construction, confirmed exposures are exposures to agents who have
  // tested positive within the simulation and the positive test result has been
  // reported. The result must be available for the source agent before the test
  // result is reported to the agent who "owns" the ExposureResult. This
  // attempts to approximate how test results are handled by Health Authorities
  // (HAs) in the real world with incomplete information. Unconfirmed exposures
  // are associated with all the source agents in the simulation who we know are
  // positive. This included agents that are tested and not yet reported as well
  // as untested.
  enum ExposureType {
    UNSPECIFIED_EXPOSURE_TYPE = 0;
    CONFIRMED = 1;
    UNCONFIRMED = 2;
  }
  message Exposure {
    google.protobuf.Timestamp exposure_time = 1;
    // Potentially empty if we do not have that information in the simulator.
    google.protobuf.Duration duration_since_symptom_onset = 2;
    google.protobuf.Duration proximity_trace_temporal_resolution = 3;
    repeated float proximity_trace = 4;
    ExposureType exposure_type = 5;
    int64 source_uuid = 6;  // For debugging. Not used in learning.
    google.protobuf.Duration duration = 7;
    float distance = 8;
    float attenuation = 9;
    // For debugging.
    float infectivity = 10;
  }
  message ExposureResult {
    int64 agent_uuid = 1;
    abesim.TestOutcome.Outcome outcome = 2;
    google.protobuf.Timestamp test_administered_time = 3;
    google.protobuf.Timestamp test_received_time = 4;
    repeated Exposure exposures = 5;
    // For simulation purposes only: infection_onset_time will not be available
    // in real world. It may not be set if test was a false positive.
    google.protobuf.Timestamp infection_onset_time = 6;
    // In certain configurations, we expose hazard for the purposes of
    // performing stratified sampling for learning. This is not a realistic
    // configuration as hazard is of course unknown.
    float hazard = 7;
  }
  repeated ExposureResult exposure_results = 1;
  // Each member exposure_results will contain exposures that occur between
  // exposure_start_time and exposure_end_time.
  google.protobuf.Timestamp exposure_start_time = 2;
  google.protobuf.Timestamp exposure_end_time = 3;
}
